docs/storage-backend/inmemorybackend.md

```diff
@@ -16,18 +16,58 @@ graph = JanusGraphFactory.build().set('storage.backend', 'inmemory').open()
 ```
 
 There are no additional configuration options for the in-memory storage
-backend. As the name suggests, this backend holds all data in memory.
+backend. As the name suggests, this backend holds all data in memory, specifically
+ in the heap space allocated to the java virtual machine running Janusgraph instance.
+
 Shutting down the graph or terminating the process that hosts the
 JanusGraph graph will irrevocably delete all data from the graph. This
 backend is local to a particular JanusGraph graph instance and cannot be
 shared across multiple JanusGraph graphs.
 
 Ideal Use Case
 --------------
+Rapid testing:
 
-The in-memory storage backend was primarily developed to simplify
+The in-memory storage backend was initially developed to simplify
 testing (for those tests that do not require persistence) and graph
-exploration. The in-memory storage backend is NOT meant for production
-use, large graphs, or high performance use cases. The in-memory storage
-backend is not performance or memory optimized. All data is stored in
-the heap space allocated to the Java virtual machine.
+exploration. Automated testing and ad-hoc prototyping remains its main purpose within Janusgraph project.
+
+Production:
+
+The initial test-only implementation was further evolved to use a more compact
+ in-memory representation, which made it suitable for production use in certain scenarios, such as 
+ Janusgraph engine being embedded into an application process, or spawned dynamically on demand, where:
+- Ease of setup/configuration/maintenance/start up is important (just run Janusgraph from distribution jar, no management of clusters for backend DBs etc) 
+- Loss of data due to _unexpected_ death of host process is acceptable (the backend provides a simple mechanism for making fast snapshots to handle _expected_ restarts)
+- Size of the graph data makes it possible to host it in a single JVM process (i.e. a few tens of Gigabytes max, unless you use a specialized JVM and hardware)
+- Higher performance is required, but no expertise/resources available to tune more complex backends. Due to its memory-only nature, in-memory backend typically performs faster than disk-based ones, in queries using simple indices
+and in graph modifications. However it is not specifically optimized for performance, and does not support advanced indexing functionality. 
+
+Limitations:
+
+- Obviously the scalability is limited to the heap size of a single JVM, and no transparent resilience to failures is offered
+- The backend offers store-level locking only, whereas a Janusgraph transaction typically changes multiple stores (e.g. vertex store and index store).
```

Comment:
It means that it can use java read-write lock to lock the individual store it modifies, for the duration of the modification, to prevent concurrent modifications on the same store. But it does not attempt to lock all the stores involved in one Janusgraph/backend transaction, and so individual store updates from parallel transactions being committed can interleave. 

So say tx1 updates the index store, but before it can update the vertex store accordingly, tx2 updates the index store with its own data. As long as this data is completely different from the tx1's, this is actually fine and allows to reduce the lock contention significantly. But if they modify the same data - you can get "ghost vertices" or "missing vertices" etc.

Note that naively implementing locking all the stores at the beginning of the transaction can lead to deadlocks unless they are locked in the same order, and can lead to high contention in case of big number of parallel transactions, because half of the stores are going to be locked all the time. Basically, implementing this correctly and efficiently would be akin to implementing a robust in-memory database engine, which is not the intent of this backend at all.